indra/llcommon/lua_function.cpp

```diff
@@ -0,0 +1,894 @@
+            }
+            catch (const std::exception& error)
+            {
+                // see above comments regarding lua_State's error status
+                initLuaState();
+                return { -1, stringize(LLError::Log::classname(error), ": ", error.what()) };
+            }
+        }
+    }
+    // pop everything
+    lua_settop(mState, 0);
+
+    // If we ran a script that loaded the fiber module, finish up with a call
+    // to fiber.run(). That allows a script to kick off some number of fibers,
+    // do some work on the main thread and then fall off the end of the script
+    // without explicitly appending a call to fiber.run(). run() ensures the
+    // rest of the fibers run to completion (or error).
+    luaL_checkstack(mState, 4, nullptr);
+    // Push _MODULES table on stack
+    luaL_findtable(mState, LUA_REGISTRYINDEX, "_MODULES", 1);
+    int index = lua_gettop(mState);
+    bool found = false;
+    // Did this chunk already require('fiber')? To find out, we must search
+    // the _MODULES table, because our require() implementation uses the
+    // pathname of the module file as the key. Push nil key to start.
+    lua_pushnil(mState);
+    while (lua_next(mState, index) != 0)
+    {
+        // key is at index -2, value at index -1
+        // "While traversing a table, do not call lua_tolstring directly on a
+        // key, unless you know that the key is actually a string. Recall that
+        // lua_tolstring changes the value at the given index; this confuses
+        // the next call to lua_next."
+        // https://www.lua.org/manual/5.1/manual.html#lua_next
+        if (lua_type(mState, -2) == LUA_TSTRING &&
+            std::filesystem::path(lua_tostdstring(mState, -2)).stem() == "fiber")
+        {
+            found = true;
+            break;
+        }
+        // pop value so key is at top for lua_next()
+        lua_pop(mState, 1);
+    }
+    if (found)
+    {
+        // okay, index -1 is a table loaded from a file 'fiber.xxx' --
+        // does it have a function named 'run'?
+        auto run_type{ lua_getfield(mState, -1, "run") };
+        if (run_type == LUA_TFUNCTION)
+        {
+            // there's a fiber.run() function sitting on the top of the stack
+            // -- call it with no arguments, discarding anything it returns
+            LL_DEBUGS("Lua") << "Calling fiber.run()" << LL_ENDL;
+            if (! checkLua(desc, lua_pcall(mState, 0, 0, 0)))
+                return { -1, mError };
+        }
+    }
+    // pop everything again
+    lua_settop(mState, 0);
+    return result;
+}
+
+LuaListener::ptr_t LuaState::getListener(lua_State* L)
+{
+    // have to use one more stack slot
+    luaL_checkstack(L, 1, nullptr);
+    LuaListener::ptr_t listener;
+    // Does this lua_State already have a LuaListener stored in the registry?
+    auto keytype{ lua_getfield(L, LUA_REGISTRYINDEX, "event.listener") };
+    llassert(keytype == LUA_TNIL || keytype == LUA_TNUMBER);
```

Comment:
It would be better to either use `llassert_always` or handle it gracefully. The latter being more enduser friendly